README.md
# nexit

[![Build Status](https://travis-ci.org/ridakk/nexit.svg?branch=master)](https://travis-ci.org/ridakk/nexit)
[![Coverage Status](https://coveralls.io/repos/github/ridakk/nexit/badge.svg?branch=master)](https://coveralls.io/github/ridakk/nexit?branch=master)

***Zero Dependency*** utility to exit NodeJs gracefully

## Install

```bash
npm install nexit
```

## API

Constructor parameters

| Parameter | Default | Description |
| ----------- | ----------- | ----------- |
| options.shutdownDelay | 30000 | shutdown delay(ms) |
| options.exitDelay | 300 | exit delay(ms) |

## Quick Start

```js
const Nexit = require('nexit');

new Nexit.Nexit();
```

or

```js
import { Nexit } from 'nexit';

new Nexit();
```

## Description

This library aims to delay process exit by a configurable amount of time to let service/traffic orchestrator to stop routing active traffic so that the process can be killed safely.

1. `SIGINT`, `SIGTERM` and `uncaughtException` signals
2. start a timer with `shutdownDelay` after catching one of the signals above.
3. fire an event with `NEXIT_SHUTDOWN` to let application know about the state
> application can use this event for its `healthz` route as below.
4. once `shutdownDelay` timer is expired, start a new timer with `exitDelay`
5. fire an event with `NEXIT_EXIT` to let application know about the state
6. at this point, application may exit on its own at any time if node event loop is fully cleared
7. once `exitDelay` timer is expired, application will be killed forcefully by `process.exit` command

## Usage

Simple `healthz` route implementation for ExpressJs.

```js
const express = require('express');
const Nexit = require('nexit');

const app = express();

const nexit = new Nexit.Nexit();
nexit.on(Nexit.NEXIT_SHUTDOWN, (err, signal) => {
  console.log(`server is shutting down signal: ${signal}`, err);
  app.set('isShuttingDown', true);
});
nexit.on(Nexit.NEXIT_EXIT, () => {
  console.log('server is exiting...');
});

app.get('/healthz', (req, res) => {
  const isShuttingDown = app.get('isShuttingDown');

  if (isShuttingDown) {
    res.status(503).send({
      message: 'shutting down',
    });
    return;
  }

  res.status(200).send({
    message: 'ok',
  });
});

app.listen(8080);
```

See `./demo` for a working example;

```bash
cd ./demo
npm install
node index.js
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.5](https://github.com/ridakk/nexit/compare/v1.0.4...v1.0.5) (2020-08-24)


### Bug Fixes

* emit error and signal to listeners ([851f6ae](https://github.com/ridakk/nexit/commit/851f6ae205d7906ac291671c28cea4165c64a5a4))

### [1.0.4](https://github.com/ridakk/nexit/compare/v1.0.3...v1.0.4) (2019-11-07)


### Bug Fixes

* emit error ([bf9ae10](https://github.com/ridakk/nexit/commit/bf9ae106705a292a75db65b5dc8c4619d53cba28))

### [1.0.3](https://github.com/ridakk/nexit/compare/v1.0.2...v1.0.3) (2019-10-28)

### [1.0.2](https://github.com/ridakk/nexit/compare/v1.0.1...v1.0.2) (2019-10-25)

### [1.0.1](https://github.com/ridakk/nexit/compare/v1.0.0...v1.0.1) (2019-10-25)

## 1.0.0 (2019-10-25)


### Bug Fixes

* implementation and tests ([5ff7863](https://github.com/ridakk/nexit/commit/5ff7863ab630d857a923e81ca342687892260da2))


index.test.ts
import { Nexit, NEXIT_EXIT, NEXIT_SHUTDOWN } from './index';

jest.useFakeTimers();

describe('Nexit', () => {
  beforeAll(() => {
    const processEvents: any = {};

    process.on = jest.fn((signal: any, cb: any): any => {
      processEvents[signal] = cb;
    });

    process.kill = jest.fn((pid, signal) => {
      processEvents[signal]();
    });

    jest.spyOn(process, 'exit').mockImplementation();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('custom parameters', async () => {
    const shutdownListener = jest.fn();
    const exitListener = jest.fn();
    const nexit = new Nexit();
    nexit.on(NEXIT_SHUTDOWN, shutdownListener);
    nexit.on(NEXIT_EXIT, exitListener);

    process.kill(process.pid, 'SIGTERM');

    jest.runAllTimers();

    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenNthCalledWith(1, expect.any(Function), 30000);
    expect(setTimeout).toHaveBeenNthCalledWith(2, expect.any(Function), 300);
    expect(shutdownListener).toHaveBeenCalledTimes(1);
    expect(exitListener).toHaveBeenCalledTimes(1);
  });

  it('default parameter values', async () => {
    const shutdownListener = jest.fn();
    const exitListener = jest.fn();
    const nexit = new Nexit({
      shutdownDelay: 10000,
      exitDelay: 100,
    });
    nexit.on(NEXIT_SHUTDOWN, shutdownListener);
    nexit.on(NEXIT_EXIT, exitListener);

    process.kill(process.pid, 'SIGTERM');

    jest.runAllTimers();

    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenNthCalledWith(1, expect.any(Function), 10000);
    expect(setTimeout).toHaveBeenNthCalledWith(2, expect.any(Function), 100);
    expect(shutdownListener).toHaveBeenCalledTimes(1);
    expect(exitListener).toHaveBeenCalledTimes(1);
  });

  it('ignore consecutive signals', async () => {
    const shutdownListener = jest.fn();
    const exitListener = jest.fn();
    const nexit = new Nexit();
    nexit.on(NEXIT_SHUTDOWN, shutdownListener);
    nexit.on(NEXIT_EXIT, exitListener);

    process.kill(process.pid, 'SIGTERM');
    process.kill(process.pid, 'SIGTERM');
    process.kill(process.pid, 'SIGTERM');

    jest.runAllTimers();

    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenNthCalledWith(1, expect.any(Function), 30000);
    expect(setTimeout).toHaveBeenNthCalledWith(2, expect.any(Function), 300);
    expect(shutdownListener).toHaveBeenCalledTimes(1);
    expect(exitListener).toHaveBeenCalledTimes(1);
  });
});


__tests__/index.test.ts
import { Nexit, NEXIT_EXIT, NEXIT_SHUTDOWN } from '../src/index';

jest.useFakeTimers();

describe('Nexit', () => {
  beforeAll(() => {
    const processEvents: any = {};

    process.on = jest.fn((signal: any, cb: any): any => {
      processEvents[signal] = cb;
    });

    process.kill = jest.fn((pid, signal) => {
      processEvents[signal]();
    });

    jest.spyOn(process, 'exit').mockImplementation();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('custom parameters', async () => {
    const shutdownListener = jest.fn();
    const exitListener = jest.fn();
    const nexit = new Nexit();
    nexit.on(NEXIT_SHUTDOWN, shutdownListener);
    nexit.on(NEXIT_EXIT, exitListener);

    process.kill(process.pid, 'SIGTERM');

    jest.runAllTimers();

    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenNthCalledWith(1, expect.any(Function), 30000);
    expect(setTimeout).toHaveBeenNthCalledWith(2, expect.any(Function), 300);
    expect(shutdownListener).toHaveBeenCalledTimes(1);
    expect(shutdownListener).toHaveBeenNthCalledWith(1, expect.any(Error), 'SIGTERM');
    expect(exitListener).toHaveBeenCalledTimes(1);
  });

  it('default parameter values [SIGTERM]', async () => {
    const shutdownListener = jest.fn();
    const exitListener = jest.fn();
    const nexit = new Nexit({
      shutdownDelay: 10000,
      exitDelay: 100,
    });
    nexit.on(NEXIT_SHUTDOWN, shutdownListener);
    nexit.on(NEXIT_EXIT, exitListener);

    process.kill(process.pid, 'SIGTERM');

    jest.runAllTimers();

    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenNthCalledWith(1, expect.any(Function), 10000);
    expect(setTimeout).toHaveBeenNthCalledWith(2, expect.any(Function), 100);
    expect(shutdownListener).toHaveBeenCalledTimes(1);
    expect(shutdownListener).toHaveBeenNthCalledWith(1, expect.any(Error), 'SIGTERM');
    expect(exitListener).toHaveBeenCalledTimes(1);
  });

  it('default parameter values [SIGINT]', async () => {
    const shutdownListener = jest.fn();
    const exitListener = jest.fn();
    const nexit = new Nexit({
      shutdownDelay: 10000,
      exitDelay: 100,
    });
    nexit.on(NEXIT_SHUTDOWN, shutdownListener);
    nexit.on(NEXIT_EXIT, exitListener);

    process.kill(process.pid, 'SIGINT');

    jest.runAllTimers();

    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenNthCalledWith(1, expect.any(Function), 10000);
    expect(setTimeout).toHaveBeenNthCalledWith(2, expect.any(Function), 100);
    expect(shutdownListener).toHaveBeenCalledTimes(1);
    expect(shutdownListener).toHaveBeenNthCalledWith(1, expect.any(Error), 'SIGINT');
    expect(exitListener).toHaveBeenCalledTimes(1);
  });

  it('ignore consecutive signals', async () => {
    const shutdownListener = jest.fn();
    const exitListener = jest.fn();
    const nexit = new Nexit();
    nexit.on(NEXIT_SHUTDOWN, shutdownListener);
    nexit.on(NEXIT_EXIT, exitListener);

    process.kill(process.pid, 'SIGTERM');
    process.kill(process.pid, 'SIGTERM');
    process.kill(process.pid, 'SIGTERM');

    jest.runAllTimers();

    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenNthCalledWith(1, expect.any(Function), 30000);
    expect(setTimeout).toHaveBeenNthCalledWith(2, expect.any(Function), 300);
    expect(shutdownListener).toHaveBeenCalledTimes(1);
    expect(shutdownListener).toHaveBeenNthCalledWith(1, expect.any(Error), 'SIGTERM');
    expect(exitListener).toHaveBeenCalledTimes(1);
  });
});


src/index.ts
import { EventEmitter } from 'events';

interface NexitOptions {
  shutdownDelay?: number;
  exitDelay?: number;
}

const NEXIT_SHUTDOWN = 'NEXIT_SHUTDOWN';
const NEXIT_EXIT = 'NEXIT_EXIT';

class Nexit extends EventEmitter {
  private isShuttingDown: boolean;
  private shutdownDelay: number;
  private exitDelay: number;
  constructor({ shutdownDelay = 30000, exitDelay = 300 }: NexitOptions = {}) {
    super();
    this.isShuttingDown = false;
    this.shutdownDelay = shutdownDelay;
    this.exitDelay = exitDelay;

    this.graceful = this.graceful.bind(this);

    this.bindHandlers();
  }

  private bindHandlers(): void {
    process.on('uncaughtException', this.handleUncaughtException.bind(this));
    process.on('SIGTERM', this.handleSignal.bind(this, 'SIGTERM'));
    process.on('SIGINT', this.handleSignal.bind(this, 'SIGINT'));
  }

  private unbindHandlers(): void {
    process.removeAllListeners('uncaughtException');
    process.removeAllListeners('SIGTERM');
    process.removeAllListeners('SIGINT');
  }

  private handleSignal(signal: string): void {
    this.graceful(signal, new Error(signal));
  }

  private handleUncaughtException(err: Error): void {
    this.graceful('uncaughtException', err);
  }

  private graceful(signal: string, err: Error): void {
    if (this.isShuttingDown) {
      return;
    }

    this.isShuttingDown = true;
    this.unbindHandlers();
    this.emit(NEXIT_SHUTDOWN, err, signal);

    setTimeout(() => {
      this.emit(NEXIT_EXIT);

      setTimeout(() => {
        // .unref is required to not to block event loop to wait until
        // this timer to expire if process is able to exit on its own.
        process.exit(1); // eslint-disable-line no-process-exit
      }, this.exitDelay).unref();

      // We can set process.exitCode and allow the process to exit
      // naturally once event loop is empty
      //
      // But need to be sure that we are avoiding scheduling any
      // additional work for the event loop otherwise will
      // always be shutting down forcefully with timer above.
      //
      // Any existing timers, open connections to dbs etc...
      // must be closed in order to perform graceful shutdown
      // properly.
      process.exitCode = 1;
    }, this.shutdownDelay);
  }
}

export default Nexit;
export { Nexit, NEXIT_SHUTDOWN, NEXIT_EXIT };
